exploration/dataflow-composability.md

```diff
@@ -0,0 +1,792 @@
+inspect the raw value and formatting options
+of their arguments.
+In addition, while a custom formatter may eagerly
+format its operand to a string,
+returning the raw underlying value
+and the formatting options used for formatting
+are also useful,
+in case another function wants to extend these options
+or use them for other logic.
+
+Making the underlying structure of its inputs,
+as well as requiring formatters to return structured outputs,
+makes it possible to specify how different functions
+can be _composed_ together
+(as shown in example 1.1 below).
+
+### Pull Requests
+
+The pull request for this design document itself is [#645](https://github.com/unicode-org/message-format-wg/pull/645).
+
+A draft pull request, [#646](https://github.com/unicode-org/message-format-wg/pull/646),
+shows what the [formatting spec](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md)
+would look like if this design document were accepted.
+(As of this writing, #646 reflects a slightly older version
+of this design document, so some of the names used are different.)
+
+## Background
+
+In the accepted version of the spec (as of this writing),
+the term "resolved value" is used for several different kinds
+of intermediate values,
+and the structure of resolved values is left completely
+implementation-specific.
+
+Providing a mechanism for custom formatters to inspect more
+detailed information about their arguments requires the
+different kinds of intermediate values to be differentiated
+from each other and more precisely specified.
+
+At the same time, the implementation can still be given freedom
+to define the underlying types for representing formattable values
+and formatted results. This proposal just defines wrappers
+for those types that implementations must use in order to
+make custom functions as flexible as possible.
+
+## Use-Cases
+
+Use cases from [issue 515](https://github.com/unicode-org/message-format-wg/issues/515):
+
+The following code fragment
+invokes the `:number` formatter on the literal `1`,
+binds the result to `$a`, and then invokes `:number` on the
+value bound to `$a`.
+
+If the value of `$a` does not allow for inspecting the previous options
+passed to the first call to `:number`,
+then the `$b` would format as `1.000`.
+
+### Example 1.1
+```
+.local $a = {1 :number minIntegerDigits=3} // formats as 001.
+.local $b = {$a :number minFractionDigits=3} // formats as 001.000
+// min integer digits are preserved from the previous call.
+```
+
+In other words: the user likely expects this code to be equivalent to:
+
+### Example 1.2
+```
+.local $b = {1 :number minIntegerDigits=3 minFractionDigits=3}
+```
+
+But without `:number` being able to access the previously passed options,
+the two fragments won't be equivalent.
+This requires `:number` to return a value that encodes
+the options that were passed in, the value that was passed in,
+and the formatted result;
+not just the formatted result.
```

Comment:
Did you consider the counter example, in which the user does *NOT* want transitivity?

```
.input $datetime = {|2024-02-14T01:23:45Z| :datetime}
.local $date = {$datetime :date}
.local $time = {$datetime :time}
```

```
.local $a = {|1.234 :number maximumFractionDigits=2 maximumDecimalDigits=0}
.local $b = {$a :integer}
{{This prints 1.23: {$b}.{$a} (Yes, its perverse)}}
```